torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -223,11 +225,41 @@ struct DeviceTypePropagationPass {
     return changed;
   }
 
-  /*
-  bool processLoop(Node* n) {
+  bool applyTensorProps(
```

Comment:
why not use https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/shape_analysis.cpp#L34 ?

Reply:
I don't want union types, the code is not built to handle union types, and I want to be able to merge some properties even if other properties are not mergable.

Reply:
> I don't want union types, the code is not built to handle union types,

What does this mean? `unifyTypes` will not result in Union types. It's unifying the tensor properties themselves.

Reply:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/type.cpp#L270

Reply:
Note the `default_to_union` argument in that code.

Reply:
Refactored to just use the ShapeAnalysis code.